assets/src/state.ts

```diff
@@ -825,18 +825,22 @@ const routeTabsPushInProgressReducer = (
 }
 
 const routeTabsToPushReducer = (
-  routeTabsToPush: RouteTab[] | null,
-  newRouteTabs: RouteTab[] | null,
+  routeTabsToPush: RouteTab[][],
+  newRouteTabs: RouteTab[],
   routeTabsUpdated: boolean,
   action: Action
-): RouteTab[] | null => {
+): RouteTab[][] => {
   switch (action.type) {
     case "STARTING_ROUTE_TABS_PUSH":
-      return null
+      return routeTabsToPush.slice(1)
     case "RETRY_ROUTE_TABS_PUSH_IF_NOT_OUTDATED":
-      return routeTabsToPush || action.payload.routeTabsToRetry
+      return routeTabsToPush.length > 0
+        ? routeTabsToPush
+        : [action.payload.routeTabsToRetry]
     default:
-      return routeTabsUpdated ? newRouteTabs : routeTabsToPush
+      return routeTabsUpdated
+        ? [...routeTabsToPush, newRouteTabs]
+        : routeTabsToPush
```

Comment:
I'm pretty confused by the queue code here. Am I understanding this right?
* When the update starts, the first batch is popped from the queue.
* If it retries, the queue doesn't change. But how does the `put` call know about the batch to try with if it's no longer in the queue because it was popped earlier?
* If it fails but shouldn't retry, it remembers the updates it didn't retry, but forgets the rest of the queue?

Then what's the `default` case? The typechecker should let you be explicit about the `action.type` there, right?

Reply:
> If it retries, the queue doesn't change. But how does the put call know about the batch to try with if it's no longer in the queue because it was popped earlier?

In that case, the `retryRouteTabsPushIfNotOutdated` action will put it back on the queue.

> If it fails but shouldn't retry, it remembers the updates it didn't retry, but forgets the rest of the queue?

If it fails and shouldn't retry, it just doesn't call `retryRouteTabsPushIfNotOutdated` and otherwise proceeds with the queue as before - the rest of the queue doesn't change.

> Then what's the default case? The typechecker should let you be explicit about the action.type there, right?

That's all of the other possible actions on the state reducer, including every possible action that might update route tabs (in which case `routeTabsUpdated` will be true), but also other actions that don't do anything to route tabs (in which case it will be false and the reducer won't change the value).